### DISCUSSION TITLE
Issue Mapping <S-Tab> in IdeaVim

### DISCUSSION THREAD

**Discussion by luanlouzada** [JetBrains/ideavim#938] [General]
Hello,

I'm trying to configure IdeaVim to map <S-Tab> (Shift + Tab) to <Esc>la in insert mode. I've added the following mapping to my .ideavimrc file:

inoremap <S-Tab> <Esc>la

However, this mapping is not working. I've verified that the .ideavimrc file is being loaded correctly and tried other key combinations that work, but it seems that IdeaVim is not recognizing <S-Tab> properly.

Has anyone encountered this issue before, and is there a known solution to make the <S-Tab> mapping work in IdeaVim? Any help would be greatly appreciated!

Thank you!

**Comment by sortraev** [comment]
Hello! Half a year late, but I am also having trouble mapping shift-tab. Specifically, I wish to map shift-tab to <Esc> when in insert mode (i.e. to exit insert mode).

Did you figure out how to do it?

**Reply by luanlouzada** [reply]
I ended up giving up. I miss it because I press it in Neovim, and it jumps to the end for me. It's useful when I want to jump over a ) or }, etc.

**Reply by sortraev** [reply]
Damn! I am currently being "forced" to move from neovim to Idea/IdeaVim for my job, and this is really breaking my workflow, haha. If you ever hear about a fix, please let me know!

**Reply by citizenmatt** [reply]
Unfortunately, `<Tab>` has a lot of uses in the IDE, such as expanding Live Templates or invoking [Emmet](https://emmet.io) support. Putting `<Tab>` through the normal IdeaVim pipeline would mean these features stop working, and not putting it through the pipeline means it's not possible to remap. So IdeaVim will always decline to handle `<Tab>` in Insert mode - it's always up to the IDE to handle. This does mean that you can remap `<Tab>` and `<S-Tab>` in other modes, such as Normal or Visual, but not Insert. 

There's an issue for this, but sadly, there are no current plans to change this, or even how to change it: [VIM-2331](https://youtrack.jetbrains.com/issue/VIM-2331/Cant-remap-Tab-key-in-insert-mode)

Out of interest, what were you trying to map `<S-Tab>` too? It's missing in the original post.

**Reply by luanlouzada** [reply]
vim.keymap.set("i", "<S-Tab>", "<Esc>la", { noremap = true, silent = true })

**Reply by citizenmatt** [reply]
Has that stripped stuff again? Wouldn't that just insert `la`? Can you put it inside backticks?

**Reply by luanlouzada** [reply]
Yes, but I need to press esc and move my hand and that's a lot of times when coding, I like having to just press shift tab within my reach, it's been mapped out in my head for years.

**Reply by sortraev** [reply]
Awesome, thanks for the explanation, @citizenmatt -- it's nice to know that I can stop looking for a work-around (outside of starting to hack out a PR, I guess).

However, I'm not quite sure I understand the motivation for passing `<Tab>`s (or any other keycodes, for that matter) on to the IDE. I have already removed/overridden numerous IDE shortcuts in favor of my vim bindings, as I assume most native vim users do when moving to IdeaVim.

This leads me to a couple of questions:
1) are you saying that sending key combinations including `<Tab>` to IdeaVim would break IDE features even if IdeaVim does not pick up on the particular key combination? 
2) you say that it would break IDE *features*, but I assume you mean it would break *shortcuts* to those features..?
3) IdeaVim adds other bindings already mapped by default in e.g. IntelliJ, such as ctrl+r or ctrl+w. Why is it not an issue with these (and other) key combinations?

Thank you very much! :)

**Reply by citizenmatt** [reply]
Tabs are special, as are Enter and Escape (and possibly a couple of other keystrokes, but they're not as critical).

IdeaVim installs a "typing handler" with the editor, so it gets access to all keys pressed. This allows it to intercept and provide different behaviour for simple keys, such as `x` in Normal mode, or typing in Insert. It can change behaviour, or map keys to other keys, or let the key through to the editor to be added to the text. 

However, shortcuts are handled before the editor is notified about a keystroke. So something like CTRL-D will invoke the IDE's action (e.g. Duplicate Line) before the editor sees the keystroke. The action can interact with the editor and do stuff. So IdeaVim also creates its own IDE action that will forward the keystroke through the Vim pipeline. We gather all Vim shortcuts that conflict with IDE actions and register them for this action directly with each editor's UI component. The IDE's action system means that we then get priority when the editor has focus, and the IDE's other actions get the chance when it doesn't.

But Tab, Enter and Escape are special. They have multiple features associated with them - hitting Tab can expand a Live Template or invoke Emmet, or various other features. Hitting Escape will cancel highlighting, cancel completion, or whatever. Enter can accept some prompt, enter a newline, handle indent and so on. But the IDE action system doesn't allow multiple actions for the same shortcut on the same component - there's no resolution of what action would get called, no concept of ordering, or chaining, etc. So the editor has implemented this itself, and has a single action for Tab, Enter and Escape, and multiple handlers for the actions that should happen when the keystroke is received. These handlers are registered with an order, and if they don't process the keystroke, it's passed to the next handler in the chain.

IdeaVim does handle Tab through its custom action and shortcuts registered to the editor. But we explicitly refuse to handle it in Insert mode (the custom action marks itself as unavailable when asked to `update` in response to Tab in Insert mode). The keystroke then gets handled by the editor, which passes it into the chain of handlers, and the appropriate editor feature is invoked. When we do handle the key, we push it through the Vim pipeline, and any maps would be processed. In Normal mode, we move to the next jump location (opposite of `<C-T>` or `<C-O>`). We still have the action for Insert mode, although it's never called - we invoke the IDE's `EditorTab` action, which would insert a tab char. I _think_ it would also invoke the handlers, but I'd need to check, and I suspect it breaks ordering - the editor would be expecting to expand templates, etc. but we're trying to insert a tab instead, and then the feature is broken (this is all related to [VIM-674](https://youtrack.jetbrains.com/issue/VIM-674/Emmet-broken-in-0.33)).

There are probably a couple of things we could do to help here. Since this was put in place, we've implemented the `sethandler` mechanism to be able to say that a shortcut is handled either by IDE or Vim _per mode_. We could perhaps have a default rule that Tab is handled by Vim for all modes apart from Insert/Select. This would at least allow the user to change the behaviour.

It would also be worth looking again at what happens when we pass the Tab keystroke on to the editor - maybe it would correctly call the handlers. And we could also add our own tab handler, like we do for Enter and Escape. We could (hopefully) register this at the "correct" point in the chain so that IDE features are invoked first, and then we get Vim processing. So Tab would still expand Live Templates and Emmet, but if there aren't any, then it would map to a Vim instruction.

**Reply by sortraev** [reply]
Thanks again for the very detailed answer! Can't say I understood everything, but I got the gist. Would love to hear if you make any relevant changes.

Anyway, now that I have you on the hook, and since you seem to be so familiar: I am looking to map Shift+Tab to exit insert mode (and go to normal mode). It's an unconventional mapping, sure, but it's been muscle memory for some 7 years. Can you think of any sort of work-around in Intellij/IdeaVim? It's possible to create IdeaVim mappings to Intellij actions, but how about the other way around?

**Reply by citizenmatt** [reply]
I'm not aware of any sure fire way to make this work. You could take a look for an action called Escape, and if it exists, try adding Shift+Tab as a shortcut. I'm away from keyboard right now, so not sure it exists, or if this would cause more trouble than it's worth.

**Reply by citizenmatt** [reply]
See JetBrains/ideavim#1272
